internal/graph/check.go

```diff
@@ -499,14 +503,21 @@ func (cc *ConcurrentChecker) checkDirect(ctx context.Context, crc currentRequest
 		dispatchChunkCountHistogram.Observe(chunkCount)
 	})
 
+	// If there are caveats on the incoming relationships, then we must require all results to be
+	// found, as we need to ensure that all caveats are used for building the final expression.
+	resultsSetting := crc.resultsSetting
+	if hasCaveats {
+		resultsSetting = v1.DispatchCheckRequest_REQUIRE_ALL_RESULTS
```

Comment:
Understanding check:

The issue is that we were short-circuiting when we found a single relation which matched the check but whose caveat was false and calling it false.

This fix says that when we have a caveat, we must find all relations that match the check so that we can evaluate the caveats on all of them and then `OR` them together so that we can decide whether the overall check was true.

The more efficient fix would be to continue evaluating relationships that match the check until we find one that also matches the caveat or we exhaust the set.

Reply:
Ah I'm reading slack and realizing that my "more efficient fix" isn't how it would work for the caching reasons described.

Reply:
Correct. We cannot just eval here because we build the *full* caveat expression and then evaluate outside of the dispatcher.